### PR TITLE
Misc updates for low rank model trainer

### PR DESCRIPTION
**Pull request by deerzq**
- added unit test for multi-label cases
- fix gradient updates
- normalizing weights
- add a simple gradient descent method
to: @hectorgon @Patrick1860

### CODE REVIEW

File: training/src/main/scala/com/airbnb/aerosolve/training/LowRankLinearTrainer.scala [hunk 1]
```diff
@@ -139,7 +153,7 @@ object LowRankLinearTrainer {
             do {
               // Pick a random other label
               val (idx, label, margin, iter) = pickRandomOtherLabel(model, labels, posIdx, posMargin, rnd, dim)
-              if (iter < dim) {
+              if (iter < dim * 2) {
```
Comment (hectorgon) [inline training/src/main/scala/com/airbnb/aerosolve/training/LowRankLinearTrainer.scala]: I think this messes up the warp loss

File: training/src/main/scala/com/airbnb/aerosolve/training/LowRankLinearTrainer.scala [hunk 2]
```diff
@@ -217,6 +230,30 @@ object LowRankLinearTrainer {
     TrainingUtils.saveModel(model, config, key + ".model_output")
   }
 
+  private def normalizeWeightVectors(model: LowRankLinearModel, maxNorm: Double) = {
```
Comment (hectorgon) [inline training/src/main/scala/com/airbnb/aerosolve/training/LowRankLinearTrainer.scala]: capWeightVector? but this should ideally call a capNorm function in FloatVector so you don't duplicate code twice

File: training/src/main/scala/com/airbnb/aerosolve/training/LowRankLinearTrainer.scala (hunk 1 above)
Comment (deerzq) [inline training/src/main/scala/com/airbnb/aerosolve/training/LowRankLinearTrainer.scala]: fixed

File: training/src/main/scala/com/airbnb/aerosolve/training/LowRankLinearTrainer.scala (hunk 2 above)
Comment (deerzq) [inline training/src/main/scala/com/airbnb/aerosolve/training/LowRankLinearTrainer.scala]: done

**Comment by deerzq** [comment]
@hectorgon removed the multilabel test since it's very flaky; make the delta_max in rprop an input parameter, since for Low rank model, we do the weight normalization, it does not make sense to allow too large step.

File: training/src/main/scala/com/airbnb/aerosolve/training/GradientUtils.scala [hunk 3]
```diff
@@ -109,7 +109,30 @@ object GradientUtils {
     })
     val stepNorms = step.map(kv => math.sqrt(kv._2.dot(kv._2)))
     log.info("Sum of Step L2 norms = " + stepNorms.sum)
+    log.info("Average of Step L2 norms = " + stepNorms.sum / stepNorms.size)
+    log.info("Max of Step L2 norms = " + stepNorms.max)
     val median = stepNorms.toBuffer.sorted.get(stepNorms.size / 2)
     log.info("Median Step L2 norms = " + median)
   }
+
+  def gradientDescent(gradients : Map[(String, String), GradientContainer],
```
Comment (hectorgon) [inline training/src/main/scala/com/airbnb/aerosolve/training/GradientUtils.scala]: This doesn't seem to be used are you keeping it just in case?

**Comment by hectorgon** [comment]
:+1:

File: training/src/main/scala/com/airbnb/aerosolve/training/GradientUtils.scala (hunk 3 above)
Comment (deerzq) [inline training/src/main/scala/com/airbnb/aerosolve/training/GradientUtils.scala]: yes, just leave it there. It might be useful in the future.

**Pull request merged**